SteamKit2/SteamKit2/Steam/WebAPI/SteamConfigurationWebAPIExtensions.cs

```diff
@@ -36,7 +37,29 @@ public static WebAPI.AsyncInterface GetAsyncWebAPIInterface(this SteamConfigurat
                 throw new ArgumentNullException(nameof(config));
             }
 
-            return new WebAPI.AsyncInterface(config.WebAPIBaseAddress, iface, config.WebAPIKey);
+            return new WebAPI.AsyncInterface(config.GetHttpClientForWebAPI(), iface, config.WebAPIKey);
+        }
+
+        internal static HttpClient GetHttpClientForWebAPI(this SteamConfiguration config)
+        {
+            var client = config.GetHttpClient();
+
+            client.BaseAddress = config.WebAPIBaseAddress;
```

Comment:
Won't this cause issues if consumer is using HttpClient for his own requests with custom base address (or lack of it)?

Reply:
I would agree here, It should be easy enough to just pass base address into `urlBuilder` directly instead of using `HttpClient.BaseAddress`.

Reply:
The xmldoc calls this out by differentiating between “returns a new or existing XXXXX” and “creates a new XXXXX”. 

Sharing an http client outside of our own library would drastically limit how much of its functionality we can actually use, both now and into the future.

Reply:
Well, SK2 itself will share HttpClient in different places (e.g with CDNClient).

Reply:
IIRC a single CDNClient has a single HttpClient, so I wouldn't call that sharing.

Reply:
It depends on general approach and preference, but I wouldn't really call it limitation. I don't see any use case in which you'd be somehow limited from doing a particular action just because you're not expected to edit behaviour of `HttpClient`. I mean, right now SK2 modifies only two of those properties, `BaseAddress` and `Timeout`, out of which one is pretty optional to use and can be achieved with url building logic, while second one is customizable already.

One `HttpClient` could be shared across several different classes and functionalities if wanted, question is if we do want it to happen. On the low level, connection pool is already shared with handler, so I don't have any argument after or against, merely a preference. I don't see a situation in which you'd want to have different `HttpClient` settings for different functions, apart from `BaseAddress` which is the reason why we don't have a single `HttpClient` already.

Maybe it'd make sense to consider this, unless I'm unaware of some specific `HttpClient` setting that would vouch against sharing a single instance. I took a look and there are only headers and buffer size right now.

Reply:
Well that's the settings we have to play with, but we could also, perhaps in future, need to wrap the user-supplied HttpMessageHandler with one or more of our own (e.g. a `DelegatingHandler` subclass), and then you _have_ to have a different HttpClient instance to hold that handler.

Also, if we shared a HttpClient, SteamKit requests could never have a longer timeout set than the rest of the user's application. We could only have a shorter one, if we (or the user) so desired.

It just feels like we'd be painting ourselves into a corner for no extra gain.

Reply:
This is a fair point, if you can see how it'd bite us in the future then I see no need to do any of that. In fact, I'd even vouch for complete removal of custom `HttpClient` and leave handler only, but sadly there are `HttpClient` settings that the consumer might want to supply, mainly default request headers for identification purposes.

Maybe it could be a good idea to allow consumer to supply only its own `HttpClientHandler` and default headers, leaving creating `HttpClient` to SK2. I don't see any other reason for one to force usage of its own `HttpClient`, and because SK2 is expected to modify its properties anyway, it'd avoid a lot of confusion (even if it's stated in the doc).

Reply:
It feels fine for me to supply message handler, default headers and value specifying that my handler is supposed to be reused in my own routines further, so it shouldn't be disposed.

Considering that there is really not much more to add towards that (I mean, you've defined everything regarding how to create and use `HttpClient` by now), I feel that it's a good tradeoff for having a guarantee that `HttpClient` created by SK2 has expected state and no issues as long as user can guarantee that his message handler is proper.

Reply:
Still, this is more like a design choice over personal preferences rather some crucial decision, so I leave that up to you, the PR is fine as it is, I just don't see a reason why consumer must create `HttpClient` and ensure it meets all SK2 requirements and can be modified, if SK2 can pretty much do the same with just 1 more argument (headers + disposing vs httpclient alone).

Reply:
If the consumer:

- Doesn't supply their own message handler factory
- Supplies a new message handler each time

Then they don't need to create a `HttpClient`. If they want more advanced control, then I'd rather give them more advanced control rather than re-inventthe wheel.

For example, they could use [`HttpClientFactory`](https://github.com/aspnet/HttpClientFactory) with few (if any) modifications.